Adds support for Connection/Hook discovery from providers

potiuk opened the pull request:
This PR extends providers discovery with the mechanism
of retrieving mapping of connections from type to hook.

Fixes #12456

Depends on #12512 so only check the last commit

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

potiuk commented on airflow/provider.yaml.schema.json:
```diff
@@ -17,6 +17,11 @@
         "type": "string"
       }
     },
+    "provider-package": {
```
@ashb @mik-laj @kaxil -> this is also for preparation to be able to use 3rd-party providers. I've added provider-package to the provider.yaml to avoid repetition and made all modules/classes relative. This way it is really safe - because we will not be able to point accidentally to a class outside of our own package.

I also added a check if provider package is properly named but only when it starts with "airflow.providers". 

Currently we only support `airflow.providers` anyway. 

It might be good to discuss now and agree future conventions for those 3rd-parties eventually

* 1) are we allowing any package to contain provider?

If not:

* 2a) should all providers live in airflow.providers
* 2b) should they live in airflow.contrib.providers (or similar)

if 2a) or 2b) - > should  the packages name following the convention:

```
   apache-airflow-providers-<PROVIDER> (with '.' replaced by -). 
```


OR maybe some other approach? WDYT?

potiuk replied:
I will also split this into two PRs - one adding "provider-package" and one adding connections, to keep it separate.

Also after looking at how fast and easy it is to extract all the "provider" features (extra links, connections etc.) and keep in  the  provider.yaml,  I think maybe even should bring the discussion about conventions for 3rd-party providers to the devllst?  This was non-goal for AIP-8 and Airflow 2.0, but it seems we are SOOOO close to it that it's worth doing the extra effort and discuss and vote the 3rd-party provider conventions showing examples from our existing providers.

The nice thing about it that we can provide some examples and clear instructions on "how to add a new 3rrd-party provider" and even a way to generate really nice documentation with examples for those. 

Seems that we already pretty much have what we need for that very soon, it just needs agreement on the devlist and we still have time to discuss it before 2.0.

WDYT?

potiuk replied:
As discussed with @mik-laj -> I will go back to absolute imports and change it to verify if they are the same as provider-package. It will be easier to maintain it.

ashb replied:
>         are we allowing any package to contain provider?

Yes, I firmly believe this should be the case. By not having third party packages live under the `airflow.` namespace, it becomes much much easier for Us and hopefully users to know what is an official provider, and thus supported by us vs not.

I think the best approach would be something like a setuptools entrypoint.

That way _any_ package can register a provider, by adding something like this to it's setup.cfg:

```ini
[options.entry_points]
apache-airflow-provider=
    x=my_company.provider.x:register_provider
```

And then in `my_company/provider/x.py` we could have something like this:

```python
def get_provider_info():
    return {
        description: "X"
    }
```

We don't need to include the package name, nor the version, as those are available from Entrypoint already.

In short: Python already has this mechanism that is well used. Let's not re-invent something different.

potiuk replied:
Thanks! 

TL;DR; I think entrypoint idea for runtime optimization is really cool, happy to incorporate it as long as we keep "developer" experience intact (or even optimized in this case as well).

> Yes, I firmly believe this should be the case. By not having third party packages live under the airflow. namespace, it becomes much much easier for Us and hopefully users to know what is an official provider, and thus supported by us vs not.

1. Do you think we should prreprevent or enforce 3rd parties from using airflow.* ? I do not think there is a way of doing it unless we sign at least some of the package cryptographically with our own case? Or is it just convention you propose here? That other packages SHOULD NOT use "airflow" as package by convention? 

> That way any package can register a provider, by adding something like this to it's setup.cfg:

[options.entry_points]
apache-airflow-provider=
    x=my_company.provider.x:register_provider

That's a cool idea, especially that rather than returning a hard-coded dictionary, we can return the very same `provider.yaml` file, as a dictionary. That will make it super-easy for anyone writing their own provider as they will have the .json.schema that will define the expected structure of the file and it will be very easy to explain it. Very standard, supported by most IDEs with auto-completion and documentation (I used it to add the current files) and it can be used to verify such provider information during provider development.

This might help to speed up the runtime discovery - One of the problems with `walk_packages` is that it is a bit slow. So not having to use it at runtime is good. And I am happy to change the mechanism for runtime for that, as long (as discussed on slack) as we solve the problem of local provider development. We need to still support testing and developing airflow providers directly from sources in checked out airflow sources. Having to install each provider as a package in order to test it sounds like a terrible idea for people who develop providers (which are most of the casual contributors to airflow). 

If for runtime we can switch to entry_points, that makes it possible also to optimize the - directly-from-sources case. This is then much easier because we can rely on packages being in the right folder and we can use simple directory walking to find all provider files, which makes it considerably faster. Paired with the runtime "entrypoint' discovery I think we can have a very optimized solution keeping all the nice properties we have now: keeping provider "meta-data" in one place, with validated json.schema yaml file in "airflow" providers. But if someone wants, they can still keep them as dict objects or json in 3rd-party providers as long as they follow the schema. I don't think we need to force anyone to yaml for that even if it is super-convenient to us and we use it also for document generation.


Unless you have other, better idea to keep the developer experience intact.

potiuk replied:
Hey @ashb - implemented all that in #12512  and this one is only left for connections. we can discuss it separately.

ashb commented on airflow/providers/apache/cassandra/provider.yaml:
```diff
@@ -33,10 +35,15 @@ integrations:
+connections:
+  - connection-type: cassandra
+    hook-class: .hooks.cassandra.CassandraHook
+    connection-id-parameter-name: cassandra_conn_id
```
I'd rather we defined these as properties in code than in a yaml file
```suggestion
connections:
  - hook-class: .hooks.cassandra.CassandraHook
```

And then:

```python
class CassandraHook(BaseHook):
   CONNECTION_TYPE="cassandra"
   CONNECTION_ID_PARAM_NAME="cassanrda_conn_id"
```

potiuk replied:
There are many other things that are not yet defined here (other PRs are addressing it) - extra links and connection in forms in javascript. If you look at those, you will understand that those cannot be as easily incorporated in the classes. 

I prefer to keep "provider delivered functionality" in one place rather than some of it in .yaml, some of this in some connection and some of this in other classes.

Also, our proposal is far from complete - it does not actually solve the problem of discovering those Hooks (unless you have not posted it. Even with the entry_points mechanism, you will have to somehow at least keep the list of hooks in each provider (there are at least 4 in the google provider). That would mean that you would have to not only keep the list of hooks but also remember to update the  "extra" information in the hooks that you want to expose- this will have to be (by its nature) split between two places. 

What was your idea here? where did you want to keep the list of connections to load?

ashb commented on airflow/providers_manager.py:
```diff
@@ -46,42 +45,134 @@ class ProvidersManager:
     """Manages all provider packages."""
 
     def __init__(self):
+        self.__log = logging.getLogger(__name__)
```
(Not important, but why so private?)

potiuk replied:
Privacy by design. I switched to the (recommended by many) mode of programming - only expose what you expect to be used and everything else private by design. That makes it much more encapsulated and makes it easier to avoid accidental uses and unintended side effects.

ashb commented on airflow/providers_manager.py:
```diff
@@ -46,42 +45,134 @@ class ProvidersManager:
+            if "deprecated" not in args[0]:
+                old_warn(*args, **kwargs)
+
+        warnings.warn = warn
```
Use the built in mechansim for this https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings

potiuk replied:
Ah nice!

potiuk committed:
Adds support for Hook discovery from providers

This PR extends providers discovery with the mechanism
of retrieving mapping of connections from type to hook.

Fixes #12456

ashb commented on airflow/providers/apache/hive/hooks/hive.py:
```diff
@@ -78,9 +78,13 @@ class HiveCliHook(BaseHook):
 
+    conn_name_attr = 'hive_cli_conn_id'
+    default_conn_name = 'hive_cli_default'
+    conn_type = 'hive_cli'
+
     def __init__(
         self,
-        hive_cli_conn_id: str = "hive_cli_default",
+        hive_cli_conn_id: str = default_conn_name,
```
We could, if we wanted, use `inspect.signature` to get the default value for the hook:

```ipython
In [3]: from airflow.providers.apache.hive.hooks.hive import HiveCliHook                                                                                                                                                                                                                                    

In [4]: import inspect                                                                                                                                                                                                                                                                                      

In [5]: inspect.signature(HiveCliHook).parameters['hive_cli_conn_id'].default                                                                                                                                                                                                                               
Out[5]: 'hive_cli_default'
```

Not needed, and might be overkill, but is just one less thing to need in the "interface"

potiuk replied:
I want to do a miniumum set of changes for this discovery to pass, we can definitely improve it in the future.

potiuk replied:
I really want to avoid creating new "Hook" interface now. My goal is to have provider hook discovery for now. Adding, standardizing documenting and polishing "common Hook API"  should be next step IMHO - possibly at the time when we introduce reading Hooks via " plugin-lilke" interface.

ashb replied:
You might not have explicitly stated it, but this is creating a new Hook interface.

I guess the main thing is this is opt-in, so existing sub-classes don't suddenly break. Good point. :+1:

ashb commented on airflow/providers/apache/hive/hooks/hive.py:
```diff
@@ -78,9 +78,13 @@ class HiveCliHook(BaseHook):
     :type  mapred_job_name: str
     """
 
+    conn_name_attr = 'hive_cli_conn_id'
```
What was your thinking for `conn_name_attr` instead of `CONN_NAME_ATTR`?

(My default would be to use the latter, as it's a common signifier that it's a constant.)

potiuk replied:
Just following "do the same as others"  simply we already had other connection class attributes lowercase (convention used in the SQL Hooks). So I just followed it.

ashb replied:
I'm not sure the convention from SQL hooks is the best one -- those are some of the oldest parts of Airflow and not very "Good Practice".

I'm okay with it if you think this way is better, but may other places in Airflow we have UPPERCASE style.

potiuk replied:
I fully agree. But I did not want to break the compatibility (potentially someone could have used those values - especially the defaults. And my goal was just to make it works for providers, not having to deal with back-compatibility :). 

There is a polish sentence I tend to use in such cases which I learned from my friend (but it has some curse-words and rhymes, so I won't translate it here verbatim, but it roughly translates to "consistent is better than perfect".

I fully agree constants *should* be UPPERCASE of course.

potiuk replied:
BTW. Happy to make a change later (maybe even we manage in 2.0 if you think it's ok to break compatibility with the names. Once it is consistent, we can do global rename.

ashb reviewed (commented):
Should we define these properties as required on BaseHook

```
    @property
    @abstractmethod
    def conn_type(self):
        ...
```

etc.

kaxil commented on airflow/cli/commands/provider_command.py:
```diff
@@ -0,0 +1,88 @@
+        {
+            'Connection type': hook_item[0],
+            'Hook class': hook_item[1][0],
+            'Hook connection attribute nme': hook_item[1][1],
```
```suggestion
            'Hook connection attribute name': hook_item[1][1],
```

kaxil commented on airflow/models/connection.py:
```diff
@@ -30,69 +31,22 @@
+cache = lru_cache(maxsize=None)
+
+
+@cache
```
```suggestion

@lru_cache(maxsize=None)
```

Since it is just at a single place

potiuk replied:
Yep

kaxil commented on airflow/models/connection.py:
```diff
@@ -30,69 +31,22 @@
+@cache
+def get_conn_type_to_hook() -> Dict[str, Tuple[str, str]]:
+    """
+    Return A map that assigns a connection type to a tuple that contains
```
```suggestion
    Returns a map that assigns a connection type to a tuple that contains
```

kaxil commented on airflow/providers_manager.py:
```diff
@@ -36,52 +39,232 @@
+                raise Exception(
+                    f"The package '{package_name}' from setuptools and "
+                    f"{provider_info_package_name} do not match. Please make sure they are"
+                    f"aligned"
```
```suggestion
                    "aligned"
```

potiuk replied:
Joined the two lines.

ashb commented on airflow/providers/microsoft/mssql/hooks/mssql.py:
```diff
@@ -70,3 +71,9 @@ def set_autocommit(
+class MsSqlHookWithOdbcHook(OdbcHook):
+    """Interact with Microsoft SQL Server via ODBC."""
+
+    conn_type = 'mssql'
```
Don't we need 


```suggestion
    conn_type = 'mssql'
    conn_name_attr = 'mssql_conn_id'
    default_conn_name = 'mssql_default'
```

potiuk replied:
I think just mssql_default is needed. Since it derives from the OdbcHook, it also gets those class-level variables. But indeed if we want to keep consistency default_conn_name should be mssql_default. the conn_name_attr should be still `odbc_conn_id` not `mssql_conn_id` because that's the name of the parameter initialized in the OdbcHoook.

But yeah - adding it here as well explicitly rather than relying the derived value is a good idea.

ashb commented on airflow/providers/microsoft/mssql/hooks/mssql.py:
```diff
@@ -36,7 +37,7 @@ def __init__(self, *args, **kwargs) -> None:
             (
                 "This class is deprecated and will be removed in Airflow 2.0.\n"
                 "pymssql is discontinued.  See https://github.com/pymssql/pymssql/issues/668.\n"
-                "Please use `airflow.providers.odbc.hooks.odbc.OdbcHook`"
```
Is this deprecation just because pymsql _was_ deprecated but is no longer?

If so could we not remove this depreaction, and not have to add the new `MsSqlHookWithOdbcHook` class?

potiuk replied:
I did not even know it's been revived. It was not clear from the warning still there - but I see that you un-deprecated the whole module in https://github.com/apache/airflow/commit/765d29ecc9fd6a3220efa0a6c4ce10848f5cbf82 - but not the class itself, that's why I added the ODBCHook. In this case, yes - we should remove the deprecation and this class is not needed any more

ashb commented on airflow/cli/commands/provider_command.py:
```diff
@@ -0,0 +1,88 @@
+    providers = ProvidersManager().providers
+    if args.provider_name in providers:
+        provider_version = providers[args.provider_name][0]
+        provider_info = providers[args.provider_name][1]
```
```suggestion
        provider_version, provider_info = providers[args.provider_name]
```

potiuk replied:
Yep

ashb commented on airflow/models/connection.py:
```diff
@@ -30,69 +31,20 @@
-# PLEASE KEEP ABOVE LIST IN ALPHABETICAL ORDER.
+
+@lru_cache(maxsize=None)
+def get_conn_type_to_hook() -> Dict[str, Tuple[str, str]]:
```
```suggestion
def get_conn_type_to_hook_mappping() -> Dict[str, Tuple[str, str]]:
```

As an English speakger `get_conn_type_to_hook` reads to me like it should be used as `get_conn_type_to_hook(self.conn_type)`, so slight tweak to the function name please.

That said, given this is a one line function that just accesses a property on a singleton object, do we even need the function?

potiuk replied:
Not any more - I think it's a remnant from previous versions where hooks were retrieved on-demand as well.

turbaszek commented on airflow/providers_manager.py:
```diff
@@ -36,52 +39,231 @@
+        sorted_dict = OrderedDict()
+        for connection_type in sorted(self._hooks_dict.keys()):
+            sorted_dict[connection_type] = self._hooks_dict[connection_type]
+        self._hooks_dict = sorted_dict
```
```suggestion
        self._hooks_dict = OrderedDict(sorted(self._hooks_dict.items()))
```

potiuk replied:
Yep.

turbaszek commented on airflow/providers_manager.py:
```diff
@@ -36,52 +39,231 @@
+        sorted_dict = OrderedDict()
+        for provider_name in sorted(self._provider_dict.keys()):
+            sorted_dict[provider_name] = self._provider_dict[provider_name]
+        self._provider_dict = sorted_dict
```
```suggestion
        self._provider_dict = OrderedDict(sorted(self._provider_dict.items()))
```

potiuk replied:
yep

ashb commented on airflow/providers_manager.py:
```diff
@@ -36,52 +39,231 @@
+            )
+            return
+
+        self._hooks_dict[conn_type] = (hook_class_name, connection_if_attribute_name)
```
```suggestion
        connection_id_attribute_name = getattr(hook_class, 'conn_name_attr')
        if not connection_id_attribute_name:
            log.warning(
                "The hook_class '%s' is missing `conn_name_attr` attribute and cannot be registered",
                hook_class,
            )
            return

        self._hooks_dict[conn_type] = (hook_class_name, connection_id_attribute_name)
```

ashb replied:
Oh, wrong PR for this.

potiuk replied:
Happy to change it here anyway.

ashb commented on airflow/providers_manager.py:
```diff
@@ -36,52 +39,231 @@
+        conn_type = getattr(hook_class, 'conn_type')
+        if not conn_type:
+            log.warning(
+                "The hook_class '%s' misses connection_type attribute and cannot be registered",
```
```suggestion
                "The hook_class '%s' is missing `connection_type` attribute and cannot be registered",
```

ashb commented on dev/provider_packages/SETUP_TEMPLATE.py.jinja2:
```diff
@@ -59,19 +59,6 @@ def do_setup(version_suffix_for_pypi=''):
-          "*.ipynb",
-{% endif %}
-            ],
-        },
```
This is needed I think, otherwise the datafiles are not included in built wheels.

potiuk replied:
Are you sure of that? Isn't the MANIFEST mechanism enough for that? 

https://github.com/apache/airflow/commit/765cbbcd76900fd0777730aaba637058b089ac95#diff-4aee44fe1bb835097b401e18b813bf5181552d25c06a701afa60f9aec4530380R19  

```
{% if PROVIDER_PACKAGE_ID == 'amazon' %}
include airflow/providers/amazon/aws/hooks/batch_waiters.json
{% elif PROVIDER_PACKAGE_ID == 'cncf.kubernetes' %}
include airflow/providers/cncf/kubernetes/example_dags/example_spark_kubernetes_spark_pi.yaml
{% elif PROVIDER_PACKAGE_ID == 'google' %}
include airflow/providers/google/cloud/example_dags/*.yaml
include airflow/providers/google/cloud/example_dags/*.sql
{% elif PROVIDER_PACKAGE_ID == 'papermill' %}
include airflow/providers/papermill/example_dags/*.ipynb
{% endif %}
```


Or does it have to be duplicated for some reason?  Is manifest only for sdist and package_data here for whl? Do I guess correctly ? 

It would contradict the documentation though - from https://setuptools.readthedocs.io/en/latest/userguide/datafiles.html it looks like MANIFEST.in allows to specify includes alone, and package_data is an optional counterpart that allows more fine-grain control. 

> Setuptools offers three ways to specify data files to be included in your packages. First, you can simply use the include_package_data keyword, e.g.:

```
setup(
    ...
    include_package_data=True
)
```
> This tells setuptools to install any data files it finds in your packages. The data files must be specified via the distutils’ MANIFEST.in file.

> If you want finer-grained control over what files are included (for example, if you have documentation files in your package directories and want to exclude them from installation), then you can also use the package_data keyword, e.g.:

Do I read that correctly that we only need includes in MANIFEST.in ? I think I checked that the selected files in `google` and `cncf.kubernetes` were added when I removed the package_data  entry. But I will double check it again.

The version with both mechanisms cause pretty confusing behaviour and took me quite some time to discover why we have provider.yaml in the google and cncf.kubernetes packages only.

potiuk replied:
Yep - it seems to work with just MANIFEST.in:

```
jarek:~/code/airflow] [test-pip-check] add-connection-discovery-to-providers+ 1 ± tar -tvzf dist/apache-airflow-providers-google-1.0.0b2.tar.gz| grep yaml
-rw-r--r-- root/root       860 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/example_dags/example_cloud_build.yaml
```

```
[jarek:~/code/airflow] [test-pip-check] add-connection-discovery-to-providers+ ± unzip -t dist/apache_airflow_providers_google-1.0.0b2-py3-none-any.whl | grep yaml
    testing: airflow/providers/google/cloud/example_dags/example_cloud_build.yaml   OK
```

potiuk replied:
```
[jarek:~/code/airflow] [test-pip-check] add-connection-discovery-to-providers+ ± unzip -t dist/apache_airflow_providers_google-1.0.0b2-py3-none-any.whl | grep sql 
    testing: airflow/providers/google/cloud/example_dags/example_bigquery_query.sql   OK
    testing: airflow/providers/google/cloud/example_dags/example_cloud_sql.py   OK
    testing: airflow/providers/google/cloud/example_dags/example_cloud_sql_query.py   OK
    testing: airflow/providers/google/cloud/example_dags/example_dataflow_sql.py   OK
    testing: airflow/providers/google/cloud/example_dags/example_mysql_to_gcs.py   OK
    testing: airflow/providers/google/cloud/example_dags/example_spanner.sql   OK
    testing: airflow/providers/google/cloud/hooks/cloud_sql.py   OK
    testing: airflow/providers/google/cloud/operators/cloud_sql.py   OK
    testing: airflow/providers/google/cloud/transfers/bigquery_to_mysql.py   OK
    testing: airflow/providers/google/cloud/transfers/mssql_to_gcs.py   OK
    testing: airflow/providers/google/cloud/transfers/mysql_to_gcs.py   OK
    testing: airflow/providers/google/cloud/transfers/sql_to_gcs.py   OK
[jarek:~/code/airflow] [test-pip-check] add-connection-discovery-to-providers+ ± tar -tvzf dist/apache-airflow-providers-google-1.0.0b2.tar.gz| grep sql          
-rw-r--r-- root/root     10832 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/example_dags/example_cloud_sql_query.py
-rw-r--r-- root/root      2471 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/example_dags/example_dataflow_sql.py
-rw-r--r-- root/root       879 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/example_dags/example_spanner.sql
-rw-r--r-- root/root      1465 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/example_dags/example_mysql_to_gcs.py
-rw-r--r-- root/root     13397 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/example_dags/example_cloud_sql.py
-rw-r--r-- root/root       816 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/example_dags/example_bigquery_query.sql
-rw-r--r-- root/root     44573 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/operators/cloud_sql.py
-rw-r--r-- root/root     44392 2020-11-27 12:15 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/hooks/cloud_sql.py
-rw-r--r-- root/root      6769 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/transfers/bigquery_to_mysql.py
-rw-r--r-- root/root      5107 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/transfers/mysql_to_gcs.py
-rw-r--r-- root/root      3093 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/transfers/mssql_to_gcs.py
-rw-r--r-- root/root     13610 2020-11-26 14:08 apache-airflow-providers-google-1.0.0b2/airflow/providers/google/cloud/transfers/sql_to_gcs.py
[jarek:~/code/airflow] [test-pip-check] add-connection-discovery-to-providers+ ± 

```

ashb replied:
I thought manifest was for sdist and package_data was for wheel. But cool.

kaxil reviewed (approved):
Needs to rebase on the master but LGTM otherwise

mik-laj commented on airflow/provider.yaml.schema.json:
```diff
@@ -167,6 +170,13 @@
         ]
       }
+    },
+    "hook-class-names": {
```
Above we have a 'hooks' section that accepts the python-module key. Should we take advantage of this?
https://github.com/apache/airflow/blob/fb23e61317c0173f2e4508b28fd966c776739659/airflow/provider.yaml.schema.json#L120-L143

potiuk replied:
I wanted to explicity call out the Hook classes not modules. I think this will be faster and better than trying to find Hooks and there might some exceptions that we would have to handle (for example Hooks derived from other Hooks). however I wanted to make some check and see how many those exceptions we have and if we find that there are not many, we could indeed try to find all the classes deriving from Base Hook in the "hook" modules specified. I will take a look.

potiuk replied:
I added a simple check now, and I think we should stay with explicitly listing the hook classes.

Just checking one amazon provider - those are Amazon's Hooks: none of them were ever "registered" in the code as "connection type to hook" nor with the customized ConnectionForm, nor in the `connection_form.js'. See below for the list of those. Similarly answering the https://github.com/apache/airflow/pull/12466#pullrequestreview-536502242  question after being a little more informed. I do not think in this case we should require all Hooks to have those extra class fields (at least not now). The use of those fields for Hooks that do not register themselves in the Core is unclear and I think we should not require the connections to have those values. 

Maybe at some point in time we want to extend the API of the BaseHook object and make it obligatory for all Hooks, but I am not sure it is something we should do now. 

IMHO - each provider should be free to choose which Hooks of that provider should be "registered" by the core for the usage in the UI and "connection to hook mapping". @ashb @mik-laj - WDYT?

Here are the Amazon hooks which do not require any special "registering" in the core (and never did):

```
'airflow.providers.amazon.aws.hooks.athena.AWSAthenaHook'
'airflow.providers.amazon.aws.hooks.athena.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.aws_dynamodb.AwsDynamoDBHook'
'airflow.providers.amazon.aws.hooks.base_aws.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.base_aws.BaseHook'
'airflow.providers.amazon.aws.hooks.batch_client.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.batch_client.AwsBatchClientHook'
'airflow.providers.amazon.aws.hooks.batch_waiters.AwsBatchClientHook'
'airflow.providers.amazon.aws.hooks.batch_waiters.AwsBatchWaitersHook'
'airflow.providers.amazon.aws.hooks.cloud_formation.AWSCloudFormationHook'
'airflow.providers.amazon.aws.hooks.cloud_formation.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.datasync.AWSDataSyncHook'
'airflow.providers.amazon.aws.hooks.datasync.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.dynamodb.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.dynamodb.AwsDynamoDBHook'
'airflow.providers.amazon.aws.hooks.ec2.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.ec2.EC2Hook'
'airflow.providers.amazon.aws.hooks.elasticache_replication_group.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.elasticache_replication_group.ElastiCacheReplicationGroupHook'
'airflow.providers.amazon.aws.hooks.emr.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.emr.EmrHook'
'airflow.providers.amazon.aws.hooks.glacier.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.glacier.GlacierHook'
'airflow.providers.amazon.aws.hooks.glue.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.glue.AwsGlueJobHook'
'airflow.providers.amazon.aws.hooks.glue_catalog.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.glue_catalog.AwsGlueCatalogHook'
'airflow.providers.amazon.aws.hooks.kinesis.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.kinesis.AwsFirehoseHook'
'airflow.providers.amazon.aws.hooks.lambda_function.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.lambda_function.AwsLambdaHook'
'airflow.providers.amazon.aws.hooks.logs.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.logs.AwsLogsHook'
'airflow.providers.amazon.aws.hooks.redshift.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.redshift.RedshiftHook'
'airflow.providers.amazon.aws.hooks.s3.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.s3.S3Hook'
'airflow.providers.amazon.aws.hooks.sagemaker.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.sagemaker.AwsLogsHook'
'airflow.providers.amazon.aws.hooks.sagemaker.S3Hook'
'airflow.providers.amazon.aws.hooks.sagemaker.SageMakerHook'
'airflow.providers.amazon.aws.hooks.secrets_manager.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.secrets_manager.SecretsManagerHook'
'airflow.providers.amazon.aws.hooks.ses.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.ses.SESHook'
'airflow.providers.amazon.aws.hooks.sns.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.sns.AwsSnsHook'
'airflow.providers.amazon.aws.hooks.sqs.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.sqs.SQSHook'
'airflow.providers.amazon.aws.hooks.step_function.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.step_function.StepFunctionHook'
```

mik-laj replied:
That sounds good to me. If we need a new connection in Web UI, we have to load the module and mark it by adding this hook to this field.

potiuk replied:
Exactly. I will add a complete documentation on "Writing your own providerr" after we merge it all, so that I do not loose time on describing things that might change.

mik-laj replied:
However, if we have a situation where some classes require a certain set of attributes and others do not, it may be a good opportunity for refacotrization. I guess this one class has two purpose now.

potiuk replied:
What's your proposal @mik-laj ? How should we refactor it ? Also remember that it builds on previously existing fields in DBApi (I followed the same convention). 

We could use Protocol (but it's Python 3.8 https://www.python.org/dev/peps/pep-0544/) - we already backport it from typing_extensions and seems like exactly what we could use. We could have Hook also derive from DiscoverableHook (for example - I have no good name for it) and have all the hooks listed there implement it.

mik-laj replied:
For me, this is good enough for now. In the future, we can divide these classes into two separate classes. We don't really need to do this yet.

However, if you are interested, I can prepare a small snippet of code that does this.

potiuk replied:
We can easily do it as a follow-up change :)

mik-laj commented on airflow/provider.yaml.schema.json:
```diff
@@ -167,6 +170,13 @@
+      "type": "array",
+      "items": {
+        "type": "string",
+        "description": "Hooks class names that provide connection types to core"
```
```suggestion
      "type": "array",
      "description": "Hooks class names that provide connection types to core"
      "items": {
        "type": "string",
```
If we describe arrays, we can use the plural in the sentence. If we describe an item, we should use singular form in the description.

potiuk replied:
What's the difference between the two in terms of UI/validation? Should we describe item or array and what happens if we do both?

mik-laj replied:
It seems to me that we should try to describe the most general fields. If we need it, we can also describe lower level elements.

From a validation perspective, these fields are ignored, but describing the list of items provides better hints in the IDE.
Before:
<img width="609" alt="Screenshot 2020-11-28 at 18 11 53" src="https://user-images.githubusercontent.com/12058428/100521656-44e2a100-31a5-11eb-9b84-1cc55bc016e9.png">

After:
<img width="639" alt="Screenshot 2020-11-28 at 18 10 02" src="https://user-images.githubusercontent.com/12058428/100521645-3a280c00-31a5-11eb-8412-efd798a608e9.png">

potiuk replied:
OK. I went ahead fixed a few typos and added some missing descriptions in #12690 as the behaviour and descriptions were not consistent in the schema.

mik-laj commented on airflow/provider.yaml.schema.json:
```diff
@@ -125,6 +127,7 @@
             "description": "Integration name. It must have a matching item in the 'integration' section of any provider."
           },
           "python-modules": {
+            "description": "Name of the module.",
```
Here we have a similar problem as we had below. Since it's an array, we should use the plural.

potiuk replied:
I fixed all the other similar typos/grammar/missing descriptions in #12690 in the original schema as the use was not relly consistent originally..

ashb commented on airflow/providers/google/cloud/hooks/bigquery.py:
```diff
@@ -67,7 +67,9 @@
-    conn_name_attr = 'gcp_conn_id'  # type: str
+    conn_name_attr = 'gcp_conn_id'
+    default_conn_name = 'google_cloud_default'
+    conn_type = 'google_cloud_platform'
```
Shouldn't this be something like `conn_type = 'big_query'`?

potiuk replied:
Possibly. I just do not want to fix all those in this PR. I simply converted what I found in the original connection.py. 

I propose that we start agreeing the approach there and standardize those, but this shoudl be a completely different PR (and if you want to propose some guidelines and standard/consistent approach for that - feel free). 

I do not want to scope-creep that change.

mik-laj replied:
We don't need a separate type for BigQuery. In Airflow 1.10, we had a few more connection types, but in Airflow 2.0, this was standardized and most connections use one connection type. In some exceptional situations we need additional configuration options and then we use a new type of connection.
See: http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow-providers-google/latest/connections/index.html
https://github.com/apache/airflow/blob/master/UPDATING.md#normalize-gcp_conn_id-for-google-cloud

ashb replied:
@potiuk :+1: 

@mik-laj If we want to be able to get a BigQuery hook to from `Connection.get_hook` we'll need a unique type won't we?

(Right now the conn type for BQ is google_cloud_platform, but that means, say, DataProc can't be google_cloud_platform)

mik-laj replied:
@ashb Since BigQuery implements DBApiHook, we should probably create a new connection type to be used by DbApiHook.

I am afraid of registering all hooks because that will also mean that they will be loaded. And loading all modules from all Google libraries into memory is an expensive operation.

ashb replied:
Oh good to know.

We'll have to come up with some approach to registereing the information for UI without importing the hook classes then.

mik-laj replied:
Related code: 
https://github.com/apache/airflow/blob/a6cf883513280de2df80989b81eb65065ec761d1/airflow/providers_manager.py#L199-L209

ashb reviewed (changes requested):
Missing changes to the Amazon provider.

ashb reviewed (changes requested):
You're missing a lot more actually:

- Yandex
- ssh
- slack
- singularity
- sftp
- sendgrid
- segment
- samba

I stopped checking after this.

potiuk committed:
fixup! Adds support for Hook discovery from providers

ashb commented on airflow/providers_manager.py:
```diff
@@ -60,38 +63,29 @@ def __new__(cls):
         via the 'apache_airflow_provider' entrypoint as a dictionary conforming to the
         'airflow/provider.yaml.schema.json' schema.
         """
-        from airflow.plugins_manager import entry_points_with_dist
```
An import in a function should never be able to cause a circular import I thought, unless this function is also called from a top level somewhere.

That's why I put the import here, not at the top level.

ashb commented on airflow/plugins_manager.py:
```diff
@@ -30,6 +30,7 @@
 import importlib_metadata
 
 from airflow import settings
+from airflow.utils.entry_points_with_dist import entry_points_with_dist
```
Module name is a little bit specific -- how about `from airflow.utils.plugins import entry_points_with_dist`?

potiuk replied:
But I think "entry_point_with_dist" is a better name. we can use it also for non-plugins (providers_manager is technically not a plugin in the sense of "Airflow Plugin"). This is a very nice a little small utility function that can be used for anything that wants to quickly and reliably retrieve entry_points, not necessarily plugins.

ashb replied:
My problem with calling the module `entry_points_with_dist`  is that it makes it a single use module -- we'd almost never be able to put anything else in here.

How about `airflow.utils.entry_points` then?

potiuk replied:
That's fine, I can change it this way, no problem.

In general, my opinon is that there is nothing wrong in single class/function per module as long as there is nothing else it is really coupled with. 

I am not sure what benefits it could bring to group independent functions in the same module, other than accidental coupling. I prefer more files than unnecessary, accidental coupling. And you know - the design and coupling emerge over time and we could also rename it later (refactorings and renames are super easy if you use IDEs).

But I agree - having both module and function the same name is bad, so if only for that reason - happy to change it.

potiuk committed:
fixup! fixup! Adds support for Hook discovery from providers
